modules/ml/src/main/java/org/apache/ignite/ml/tree/data/TreeDataIndex.java

```diff
@@ -0,0 +1,179 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.ml.tree.data;
+
+import org.apache.ignite.ml.tree.TreeFilter;
+
+/**
+ * Index for representing sorted dataset rows for each features.
+ * It may be reused while decision tree learning at several levels through filter method.
+ */
+public class TreeDataIndex {
+    /** Index containing IDs of rows as if they is sorted by feature values. */
+    private final int[][] index;
+    /** Original features table. */
+    private final double[][] features;
+    /** Original labels. */
+    private final double[] labels;
+
+    /**
+     * Constructs an instance of TreeDataIndex.
+     *
+     * @param features Features.
+     * @param labels Labels.
+     */
+    public TreeDataIndex(double[][] features, double[] labels) {
+        this.features = features;
+        this.labels = labels;
+
+        int rows = features.length;
+        int cols = features.length == 0 ? 0 : features[0].length;
+
+        double[][] featuresCopy = new double[rows][cols];
+        index = new int[rows][cols];
+        for (int row = 0; row < rows; row++) {
+            for (int col = 0; col < cols; col++) {
+                index[row][col] = row;
+                featuresCopy[row][col] = features[row][col];
+            }
+        }
+
+        for (int col = 0; col < cols; col++)
+            sortIndex(featuresCopy, col, 0, rows - 1);
+    }
+
+    /**
+     * Constructs an instance of TreeDataIndex
+     *
+     * @param indexProj Index projection.
+     * @param features Features.
+     * @param labels Labels.
+     */
+    private TreeDataIndex(int[][] indexProj, double[][] features, double[] labels) {
+        this.index = indexProj;
+        this.features = features;
+        this.labels = labels;
+    }
+
+    /**
+     * Returns label for kth order statistic for target feature.
+     *
+     * @param k K.
+     * @param featureId Feature id.
+     */
+    public double labelInSortedOrder(int k, int featureId) {
+        return labels[index[k][featureId]];
+    }
+
+    /**
+     * Returns vector of original features for kth order statistic for target feature.
+     *
+     * @param k K.
+     * @param featureId Feature id.
+     */
+    public double[] featuresInSortedOrder(int k, int featureId) {
+        return features[index[k][featureId]];
+    }
+
+    /**
+     * Returns feature value for kth order statistic for target feature.
+     *
+     * @param k K.
+     * @param featureId Feature id.
+     */
+    public double featureInSortedOrder(int k, int featureId) {
+        return featuresInSortedOrder(k, featureId)[featureId];
+    }
+
+    /**
+     * Creates projection of current index in according to {@link TreeFilter}.
+     *
+     * @param filter Filter.
```

Comment:
No `@return` javadoc here and in methods above. Maybe it's fine, but looks strange.